include/cuco/detail/equal_wrapper.cuh

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cuco/detail/bitwise_compare.cuh>
+
+#include <cstddef>
+
+namespace cuco {
+namespace experimental {
+namespace detail {
+
+/**
+ * @brief Enum of equality comparison results.
+ */
+enum class equal_result : int32_t { UNEQUAL = 0, EMPTY = 1, EQUAL = 2 };
+
+/**
+ * @brief Equality wrapper.
+ *
+ * User-provided equality binary callable cannot be used to compared against sentinel value.
+ *
+ * @tparam T Right-hand side Element type
+ * @tparam Equal Type of user-provided equality binary callable
+ */
+template <typename T, typename Equal>
+struct equal_wrapper {
+  T sentinel_;   ///< Sentinel value
+  Equal equal_;  ///< Custom equality callable
+
+  /**
+   * @brief Equality wrapper ctor.
+   *
+   * @param sentinel Sentinel value
+   * @param equal Equality binary callable
+   */
+  __host__ __device__ constexpr equal_wrapper(T sentinel, Equal const& equal)
```

Comment:
> Note to future us: We could add a static_assert that checks if the user-provided equal is indeed a binary callable. This would help with diagnostics.

I tried to add more `static_assert` like this around and @jrhemstad 's point is that improper callable would fail at build time anyway.

Reply:
Yeah, it's a thin line between adding helpful diagnostics and being too verbose. I see Jake's point. It should still fail with a helpful error message, so there's no point for adding a custom check.